### ISSUE TITLE
Builder.from(abstract class) is not generated if abstract class is package-private and implementation class is public

### ISSUE THREAD

**Issue by avakhrenev** [immutables/immutables#236]
Seems like regression in 2.1.x

Consider the following definition:
```java
@Value.Immutable
@Value.Style(
  visibility = Value.Style.ImplementationVisibility.PUBLIC
)
/*package-private*/ abstract class Data {
  public abstract List<String> values();
  public ImmutableData append(ImmutableData that) {
    return ImmutableData.builder()
      .from(this) //error: Data cannot be converted to ImmutableData
      .addAllValues(that.values())
      .build();
  }
}
```
As of 2.1.5 it doesn't compile with error outlined in the comment above.
With immutables 2.0.18 it compiles as expected.

**Comment by elucash** [comment]
Hi. At first I thought that it is indeed a regression. But then, if was more like a fix. If You have public immutable and builder class, then clients from outside a package cannot call `from` method as it refers to package private class. If you make `Data` public, then it would resolve the problem. On the other hand, you might change the code to include cast
.
```java
@Value.Immutable
@Value.Style(visibility = Value.Style.ImplementationVisibility.PUBLIC)
/*package-private*/ abstract class Data {
  public abstract List<String> values();
  public ImmutableData append(ImmutableData that) {
    return ImmutableData.builder()
      .from((ImmutableBuildFromRegression) this)
      .addAllValues(that.values())
      .build();
  }
}
```
If we can fix it, then how? Is there any ideas how we can resolve this?

**Comment by elucash** [comment]
Just though that we can generate package-private overload of from method which takes package private abstract class

**Comment by avakhrenev** [comment]
Wow, that was fast!

Why can't they use it? Clients outside of the package don't have direct acces to the abstract `Data` class, but it doesn't prevent them to invoke `from(Data)` method with an instance of `ImmutableData` implementation class. Could you explain, why there are two methods required instead of one with base class parameter?

I don't think that explicit cast is a good workaround, making abstract class public is preferred IMHO. In our codebase making abstract class package-private serves the only purpose to avoid namespace pollution: only immutable classes are available outside.

Thank you very much for immutables: it's a pleasure to use it!

**Comment by elucash** [comment]
Here's hot it will be now with the recent change

```java
@Value.Immutable
@Value.Style(
  visibility = Value.Style.ImplementationVisibility.PUBLIC
)
/*package-private*/ abstract class Data {
  public abstract List<String> values();
  public ImmutableData append(ImmutableData that) {
    return ImmutableData.builder()
      .from(this)
      .addAllValues(that.values())
      .build();
  }
}
...
// builder have following from method signatures

// could be called outside of a package
public final Builder from(ImmutableData instance)

// can be used inside package as Data is package-private anyway
/*package-private*/ final Builder from(Data instance)
```

Will be available in the next patch release (in a couple of days). In a meantime you check the latest snapshots or build from sources. Thank you for the bug report!

**Comment by elucash** [comment]
Released in 2.1.8

**Comment by avakhrenev** [comment]
Thank you, it works!

However I've found somewhat similar issue: if abstract class is package-private, then generated `copyOf` method takes immutable implementation instead of abstract class:
```java
/**
 * Creates an immutable copy of a {@link Data} value.
 * Uses accessors to get values to initialize the new immutable instance.
 * If an instance is already immutable, it is returned as is.
 * @param instance The instance to copy
 * @return A copied immutable Data instance
 */
public static ImmutableData copyOf(ImmutableData instance) {
  if (instance instanceof ImmutableData) {
    return (ImmutableData) instance;
  }
  return ImmutableData.builder()
      .from(instance)
      .build();
}
```

Should I report it as a separate issue?

**Comment by elucash** [comment]
Right, it's a similar issue, we can do the same, generate package-private overload taking package-private abstract value type, this would be reasonable addition/promote consitency. We might reuse this issue, it's still open. Thank you for spotting this!

**Comment by elucash** [comment]
released in 2.1.9
